tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
 	}
 }
 
+/*
+ * Check if the join fromList belongs to a INNER JOIN / LEFT JOIN
+ * on two tables. A INNER JOIN / LEFT JOIN on two tables has one JoinExpr
+ * entry in fromList.
+ */
+static bool
+is_inner_or_left_join_two_tables(List *fromList)
+{
+	if (!IsA(list_nth(fromList, 0), JoinExpr) || (list_length(fromList) != 1))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the join fromList belongs to an implicit join (e.g.,
+ * SELECT * FROM metric m1, metric_name m2 WHERE m1.id=m2.id). This type
+ * of join has two RANGETBLREF entries in the fromList.
+ */
+static bool
+is_implicit_join_two_tables(List *fromList)
+{
+	if (list_length(fromList) != 2)
+		return false;
+
+	/* At the moment, the code is called only in the reference table join
+	 * pushdown path. Earlier checks ensure that both parameters are
+	 * RangeTblRefs. Recheck the condition here and handle the case properly.
+	 */
+	if (!IsA(list_nth(fromList, 0), RangeTblRef))
+		return false;
+
+	if (!IsA(list_nth(fromList, 1), RangeTblRef))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the query performs a join between a hypertble (outer) and a reference
+ * table (inner) and the join type is a LEFT JOIN, an INNER JOIN, or an implicit
+ * join.
+ */
+static bool
+is_safe_to_pushdown_reftable_join(PlannerInfo *root, List *join_reference_tables,
+								  RangeTblEntry *innertableref, JoinType jointype)
+{
+	Query *query = root->parse;
+	List *fromList;
+
+	Assert(root != NULL);
+	Assert(innertableref != NULL);
+
+	/*
+	 * We support pushing down of INNER and LEFT joins only.
+	 *
+	 * Constructing queries representing partitioned OUTER, SEMI and ANTI
+	 * joins is hard, hence not considered right now.
+	 */
+	if (jointype != JOIN_INNER && jointype != JOIN_LEFT)
+		return false;
+
+	/* Check that at least one reference table is defined. */
+	if (join_reference_tables == NIL)
+		return false;
+
+	/* Only queries with two tables and one join are supported. */
+	fromList = query->jointree->fromlist;
+	if (!is_inner_or_left_join_two_tables(fromList) && !is_implicit_join_two_tables(fromList))
```

Comment:
I have a bad feeling about using the raw parse tree at this stage of the planning. Generally we look at RelOptInfos. If you want to limit it to two participating tables, can you check the relids of the innerrel and outerrel?

Reply:
Could you elaborate on this? I would like to understand why this approach is bad. We use the fromlist also in [other places](https://github.com/timescale/timescaledb/blob/484a4ea3fc5acb58468f19d445b38c7aaf38fb7b/tsl/src/continuous_aggs/create.c#L1223) to detect the join type and the involved tables.

Reply:
The parse tree (`Query`) is not normally used at this point by Postgres. If you look at other join planning code like `make_one_rel`, it only uses the RelOptInfos. I'd say most Path-generation code uses RelOptInfos, which we could roughly call a "relational algebra" representation. If you are using the parse tree at the later stages of planning, you'll have to redo a lot of work done earlier. For example, here you have to recursively traverse the fromlist to find the joins written as subqueries. This was already done for you by `pull_up_subqueries`. Or if you want to examine the join clauses, you'll have to redo the work of `distribute_quals_to_rels`. 

Sometimes using the parse tree is unavoidable because we don't have the suitable hooks. But here it's not the case.

Not familiar with the caggs code, maybe it's the hook problem, or maybe they're just doing it wrong :)

Reply:
Thank you for the detailed explanation. In the actual join path generation (e.g., in the function `data_node_generate_pushdown_join_paths`) the `RelOptInfo` data structures are used as proposed. Also here in the `is_safe_to_pushdown_reftable_join` function, I use the data structures of the planner (i.e., `JoinType`) wherever I can. As you mentioned, a lot of work is applied to generate these data structures and they distinguish from the raw query (e.g., subqueries are pulled up). 

However, the current join pushdown implementation should only be applied if a query contains one join and two tables. My intention here is to access the raw query to check how the query looked like before it was pre-processed to get this information (the `root->simple_rel_array` contains a lot of additional `RelOptInfo`, which makes it harder to get this information from there).

Reply:
> However, the current join pushdown implementation should only be applied if a query contains one join and two tables. 

Does anything break if the query contains other joins? I'd say it shouldn't, but not sure. For the particular join you're pushing down, you can check that it's two tables with  `bms_num_members(jonrel->relids) == 2`.

Reply:
When you perform a query like:

```sql
 SELECT * FROM metric m1, metric m2, metric_name mn WHERE m1.id = mn.id AND m2.id = mn.id;
```

(`metric` is a DHT, `metric_name` a reference_table)

PostgreSQL could perform one join on all three tables or join two tables and join this result with the remaining table. In the latter case, you end up with `joinrels` where the condition `bms_num_members(joinrel->relids) == 2` holds but the join should not be pushed down.

My intention is here to detect queries where more than two tables are involved and don't push down the join in this case. This condition could be lifted in a further PR. However, we decided to use a strict filter here and remove these restrictions in the future when we know that the first join pushdown implementation works.

Reply:
> In the latter case, you end up with joinrels where the condition bms_num_members(joinrel->relids) == 2 holds but the join should not be pushed down.

Yes, we should also check that outer table is distributed, and inner table is a reference table. These three conditions should be sufficient, no? And we can check all of them without using the parse tree.

Reply:
I just tested it. This is not sufficient to check. If you execute a join as described [here](https://github.com/timescale/timescaledb/pull/5212#discussion_r1105469959), PostgreSQL generates multiple ways of performing the join. 

The first step could be the join between `metric m1` and `metric_name mn` and then a further join between this intermediate result and `metric m2` is performed. For the first join `bms_num_members(joinrel->relids) == 2` is true and we have a distributed outer table and an inner reference table. However, we don't want to push down the join in this case.

Reply:
> However, we don't want to push down the join in this case.

But why? `m1` and `m2` are used symmetrically here, this distinction looks arbitrary.

Reply:
> But why? m1 and m2 are used symmetrically here, this distinction looks arbitrary.

This would be very weird behavior to have, we would favor some of the grammatically implicit joins in the join order, based on their relative position in the query? That's not normally done in SQL.

Reply:
Even the explicit inner joins don't enforce the particular order, you have to set a special guc for it: https://www.postgresql.org/docs/current/explicit-joins.html

Reply:
I restructured my code. Now it is enough to check `bms_num_members(root->all_baserels)` at this place; accessing the parse tree is no longer needed.